056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+The user can follow the procedure described [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) in the official documentation, in order to meet this prerequisite.
+
+Also, the minimum Kafka version should be 3.5.0 with `inter.broker.protocol.version` set to “3.5”.
+The user could need to upgrade the Kafka version on its own cluster before running a migration.
+
+### Migration
+
+The migration process is semi-automatic which means having the user driving the process by interacting with the operator (mostly via annotations and custom resources creation/deletion) which is in charge to do the rest (i.e. nodes re-configuration, rolling, ...).
+
+> NOTE: this kind of approach is already in place for the Kafka version upgrade.
+
+Assume the following prerequisites, with a ZooKeeper based Kafka cluster defined by:
+
+* A `Kafka` custom resource.
+* One or more `KafkaNodePool`(s) for running the ZooKeeper based brokers.
+
+It would assume the operator is not using KRaft by default nor has the `+UseKRaft` feature gate enabled.
+
+First step for the user would be enabling the `+UseKRaft` feature gate unless it’s already enabled by default in a new operator release (depending on which Strimzi version this proposal will land).
+So it could be needed doing an upgrade of the current operator to the new release.
+In both cases, the operator is rolled, a new one starts and it’s able to handle the creation of new Apache Kafka clusters in KRaft mode but not migrating the old ZooKeeper based ones automatically.
+
+Currently, the operator prints the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+This is because with KRaft mode enabled in the operator, the reconciled `Kafka` resource doesn’t have any controller nodes (in any pool).
+
+The migration process can be also rolled back to ZooKeeper at any point before taking the KRaft controllers out of the migration mode.
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `MigrationManager` component in charge of tracking the migration finite state machine.
+Transitioning across the states depends on the current state, stored in the `Kafka.status.migrationState` field, and the "input" represented by the user interaction with the operator through the following annotations:
+
+* `strimzi.io/kraft-migration`: to be applied on the `Kafka` custom resource with `true|false` values to start/end the overall migration process.
+* `strimzi.io/kraft-migration`: to be applied on the `KafkaNodePool`(s) custom resource with `dual|kraft|zk` values to represent the the desired "direction" in the migration process by the user:
```

Comment:
What I was looking for the annotation on the node pools were values to set the "direction" of the migration flow more than "state"(s). I came with these three values which tbh I didn't like a lot anyway (i.e. what the user would understand about "dual"? :-) It's a kind of implementation details of the migration process itself being in a "dual write" state).
Because of having in mind steering the "direction" I didn't use what `ZkMigrationState` enums represent which are actual states. At the same time, are those values understandable to the user? (at least more than dual, zk, kraft?). Maybe documenting them we could make the meaning to be clear.

Going through their current meaning we could have:

* `pre-migration`: the Kafka meaning is controllers created, quorum reached, ZK migration enabled. If taken as annotation value it would make the state machine transition from `WaitControllersInMigration` to `WaitBrokersInMigration`. We could assume it as the `PRE_MIGRATION` in the Kafka codebase.
* `migration`: the Kafka meaning is migration done and controllers in "dual write". Of course, taken as annotation value it would mean "start the migration" (not migration already done), so moving from `WaitBrokersInMigration` to `DualWrite/WaitBrokersOutMigration` through the ephemeral `Migration` state. The final `DualWrite/WaitBrokersOutMigration` would match the `MIGRATION` meaning in the Kafka codebase.
* `post-migration`: the Kafka meaning is migration completed. Used as annotation value it would make the transition from `DualWrite/WaitBrokersOutMigration` to `WaitControllersOutMigration` and from this one to `FullKRaft`. The last `FullKRaft` would have the same meaning as the `POST_MIGRATION` in the Kafka codebase. Remember that at this point we need to delete ZooKeeper which I think is not part of the `POST_MIGRATION` meaning.

Finally, I think that the `pre-migration` could also replace `zk` to rollback the migration process where possible (as per the diagram). It doesn't have a conflict with the `pre-migration` used to move forward at the beginning of the process.

@fvaleri do you find the above description accurate and what you had in mind? If yes, I can change to use those annotation values but not representing the "state" but still the "direction", the "desire", the "next move" in the migration process. I.e. `pre-migration` = move to be ready for migration with controllers running and quorum formed connected to ZK, `migration` = run the migration by configuring brokers for ZK migration, `post-migration` = finilize the migration exiting controllers first and brokers later to be connected to ZK. Use `pre-migration` = come back to controllers running, quorum formed, connected to ZK but brokers back to use ZK.

Reply:
I think we agree on the 3 migration states/directions. I see the annotation's value as the desired state, which you can also read as the direction in which we are going to reconcile. Both interpretations are fine for me.

If we also consider @scholzj suggestion of using just one annotation on the Kafka resource, the migration process would be much simpler from the administrator perspective.

1. **kraf-migration: pre-migration**: Wait for the dedicated controller and broker pools for this cluster, without doing any rolling update.
2. **kraf-migration: migration**: Trigger the controller and broker pool migration configuration changes, and do the 1st rolling update (controller first, then brokers). This starts the cluster metadata migration, with the active controller running in dual-write mode (rollback is still possible).
3. **kraf-migration: post-migration**: As soon as the metadata migration is completed, finalize by removing the migration configuration from all pools, and doing the 2nd rolling update (rollback is not possible).

Internally, the FSM will have more states to handle the migration steps ordering (this ius as you designed). Both controller and broker migration settings can be applied automatically by the Cluster Operator once the pools are ready, but we need to roll controller first and then brokers.

When the administrator removes the Kafka annotation, and we are not in `post-migration` state, the Cluster Operator will revert back to `pre-migration` state by following the reverse path, otherwise nothing will happen. 

When in `post-migration` state, ZooKeeper is not used anymore, so the administrator will be able to safely remove its configuration from the Kafka resource, which will terminate all related pods.

Reply:
Using just `Kafka` was my first brainstormed idea but I put it aside to give more fine control to the user on the node pools (because anyway it's a matter of configuring & rolling the nodes in different pools at different times).
My thought was also having more user involved because this process will be temporary so removed in the future when all Kafka cluster will be migrated to KRaft.
Anyway the current solution could have its challenges, let me try to get back what I had before by also using some of your feedback.

> kraf-migration: pre-migration: Wait for the dedicated controller and broker pools for this cluster, without doing any rolling update.

Actually here we are waiting only for dedicated controllers which need to be configured with ZK migration enabled already. We are not waiting for broker pools. Brokers are already running using ZK and are already part of a pool. Applying `kraf-migration: pre-migration` would just mean moving to `WaitControllersInMigration` to wait the user creating the `KafkaNodePool` for controllers. When the pool is created, controllers are rolled and finally we move in `WaitBrokersInMigration`. We are still in an overall PRE_MIGRATION state (as per Kafka codebase).

> kraf-migration: migration: Trigger the 1st rolling update (controllers first, then brokers). This starts the cluster metadata migration, with the active controller running in dual-write mode (rollback is still possible).

Related to the previous point you don't need to roll controllers. When deployed they are already configured by the operator with ZK migration enabled. Applying `kraft-migration: migration` would mean configure brokers with ZK migration enabled (also connected to controllers), rolling them ... this would start the actual migration (when brokers join the controllers) so moving to `Migration` and then `DualWrite` state. We would be in an overall MIGRATION state (as per Kafka codebase). From here user can rollback to apply `kraf-migration: pre-migration` again (back to the `WaitBrokersInMigration`).

> kraf-migration: post-migration: As soon as the metadata migration is completed, finalize by removing the migration configuration from all pools, and doing the 2nd rolling update (rollback is not possible).

At this point, with this approach we are missing a potential point of rollback. My proposal has the user involved to move brokers out first and then controllers. When brokers are out (we moved to `WaitControllersOutMigration`) we can still rollback. While with this "new" approach the rollback is possible only in the `DualWrite`. Applying the `kraf-migration: post-migration` would mean moving brokers out of migration then controllers as well, all in one transition. Migration ends with the `FullKRaft` state. We are in the overall POST_MIGRATION state (as per Kafka codebase).

> When in post-migration state, ZooKeeper is not used anymore, so the administrator will be able to safely remove its configuration from the Kafka resource, which will terminate all related pods.

Currently it's not possible to remove the `Kafka.spec.zookeeper` field but we would need something like this to proceed with the ZooKeeper pods deletion.

Let me draw a new FSM I will post here before moving to have it in the proposal.

Reply:
Here a new FSM ...

![zk-kraft-migration-fsm](https://github.com/strimzi/proposals/assets/5842311/4efdf523-428a-4744-9a2c-ff7b683067ae)

Reply:
LGTM. Thanks.

Reply:
Proposal updated. I moved the previous FSM to the rejected alternatives.